File: backend/models/userModel.js
const mongoose = require('mongoose');
const { Schema } = mongoose;

// // 定義補給站領取紀錄schema
// const supplyLogSchema = new Schema({
//   spotId: {
//     type: String, // 字串存 supplyId(key)
//     required: true
//   },
//   nextClaimTime: {
//     type: Date,
//     required: true
//   }
// }, { _id: false }); // 不要自動生成 _id

// 定義使用者模型結構
const userSchema = new Schema({
  email: {
    type: String,
    required: true,
    unique: true
  },
  displayName: {
    type: String,
    default: ""
  },
  photoURL: {
    type: String,
    default: ""
  },
  role: {
    type: String,
    default: "player"
  },
  createdAt: {
    type: Date,
    default: Date.now
  },
  lastLogin: {
    type: Date,
    default: Date.now
  },
  gender: {
    type: String,
    enum: ['男', '女', '不透露'],
    default: '不透露'
  },
  age: {
    type: String,
    default: "18"
  },
  username: {
    type: String,
    default: ""
  },
  backpackItems: [{
    _id: false, //不會自動生成 _id
    itemId: {
      type: mongoose.Schema.Types.ObjectId,
      required: true
    },
    quantity: {
      type: Number,
      required: true,
      min: 1
    }
  }],
  missions: [{
    _id: false,
    taskId: {
      type: mongoose.Schema.Types.ObjectId,
      required: true
    },
    state: {
      type: String,
      enum: ['available', 'in_progress', 'completed', 'claimed', 'declined', 'deleted'],
      required: true
    },
    acceptedAt: {
      type: Date,
      default: null
    },
    expiresAt: {
      type: Date,
      default: null
    },
    refreshedAt: { 
      type: Date,
      default: null
    },
    haveCheckPlaces: [{
      _id: false,
      spotId: {
        type: Schema.Types.ObjectId,
        ref: 'Spot',
        required: true
      },
      isCheck: {
        type: Boolean,
        default: false
      }
    }],
    isLLM: {
      type: Boolean,
      required: false
    }
  }],
  spotsScanLogs: {
    type: Map,
    of: Boolean,
    default: {}
  },
  supplyScanLogs: {
    type: Map,
    of: Date,
    default: {}
  },
  settings: {
      music: {
        type: Boolean,
        default: true
      },
      notification: {
        type: Boolean,
        default: true
      },
      language: {
        type: String,
        default: 'zh-TW'
      }
    },
  buff: {
    type: Object,
    default: null
  }
}, { collection: 'users' });

// 明確指定集合名稱為 'users'
const User = mongoose.model('User', userSchema);

module.exports = User;

File: frontend/app/src/main/java/com/ntou01157/hunter/models/model_api/UserModel.kt
package com.ntou01157.hunter.models.model_api

import com.google.gson.annotations.SerializedName
import java.util.Date

// 用戶模型
data class User(
    @SerializedName("_id") val id: String,
    @SerializedName("username") val username: String,
    @SerializedName("backpackItems") val backpackItems: List<BackpackItem>,
    @SerializedName("missions") val missions: List<Mission> = emptyList(),
    @SerializedName("gender") val gender: String?,
    @SerializedName("age") val age: String?,
    @SerializedName("photoURL") val photoURL: String? = null,
    @SerializedName("settings") val settings: Settings?,
    @SerializedName("spotsScanLogs") val spotsScanLogs: Map<String, Boolean>? = null,
    @SerializedName("buff") val buff: List<Buff>? = null

    // 其他用戶屬性...
)

data class SpotUpdateBody(@SerializedName("spotId") val spotId: String)

// 背包中的物品
data class BackpackItem(
    @SerializedName("itemId") val itemId: String,
    @SerializedName("quantity") val quantity: Int
)

// 使用者任務進度模型
data class Mission(
    @SerializedName("taskId") val taskId: String,
    @SerializedName("state") val state: String,
    @SerializedName("acceptedAt") val acceptedAt: Date?,
    @SerializedName("expiresAt") val expiresAt: Date?,
    @SerializedName("refreshedAt") val refreshedAt: Date?,
    @SerializedName("haveCheckPlaces") val haveCheckPlaces: List<HaveCheckPlaces> = emptyList(),
    @SerializedName("isLLM") val isLLM: Boolean = false
)

// 任務打卡點狀態
data class HaveCheckPlaces(
    @SerializedName("spotId") val spotId: String,
    @SerializedName("isCheck") val isCheck: Boolean
)

// 使用者設定模型
data class Settings(
    @SerializedName("music") val music: Boolean,
    @SerializedName("notification") val notification: Boolean,
    @SerializedName("language") val language: String
)

data class Buff(
    @SerializedName("name") val name: String,                       // e.g. "ancient_branch"
    @SerializedName("expiresAt") val expiresAt: String              // e.g. "2025-08-24T22:46:28.153Z"
) {
    fun expiresAtMillisOrNull(): Long? =
        runCatching { java.time.Instant.parse(expiresAt).toEpochMilli() }.getOrNull()
}

fun List<Buff>?.expireAtOf(name: String): Long? =
    this?.firstOrNull { it.name == name }?.expiresAtMillisOrNull()